internal/workflow/activities/poll_sip_statuses.go

```diff
@@ -0,0 +1,116 @@
+package activities
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"time"
+
+	"github.com/google/uuid"
+	"go.artefactual.dev/tools/ref"
+	temporal_tools "go.artefactual.dev/tools/temporal"
+
+	goaingest "github.com/artefactual-sdps/enduro/internal/api/gen/ingest"
+	"github.com/artefactual-sdps/enduro/internal/enums"
+	"github.com/artefactual-sdps/enduro/internal/ingest"
+)
+
+const PollSIPStatusesActivityName = "poll-sip-statuses-activity"
+
+type PollSIPStatusesActivityParams struct {
+	BatchUUID        uuid.UUID
+	ExpectedSIPCount int
+	ExpectedStatus   enums.SIPStatus
+}
+
+type PollSIPStatusesActivityResult struct {
+	AllExpectedStatus bool
+}
+
+type PollSIPStatusesActivity struct {
+	ingestsvc    ingest.Service
+	pollInterval time.Duration
+}
+
+func NewPollSIPStatusesActivity(ingestsvc ingest.Service, pollInterval time.Duration) *PollSIPStatusesActivity {
+	return &PollSIPStatusesActivity{
+		ingestsvc:    ingestsvc,
+		pollInterval: pollInterval,
+	}
+}
+
+func (a *PollSIPStatusesActivity) Execute(
+	ctx context.Context,
+	params *PollSIPStatusesActivityParams,
+) (*PollSIPStatusesActivityResult, error) {
+	h := temporal_tools.StartAutoHeartbeat(ctx)
+	defer h.Stop()
+
+	ticker := time.NewTicker(a.pollInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-ticker.C:
+			result, err := a.checkSIPStatuses(ctx, params.BatchUUID, params.ExpectedSIPCount, params.ExpectedStatus)
+			if err != nil {
+				return nil, fmt.Errorf("check SIP statuses: %v", err)
+			}
+			if result.done {
+				return &PollSIPStatusesActivityResult{AllExpectedStatus: result.allExpected}, nil
+			}
+		}
+	}
+}
+
+type checkResult struct {
+	done        bool
+	allExpected bool
+}
+
+func (a *PollSIPStatusesActivity) checkSIPStatuses(
+	ctx context.Context,
+	batchUUID uuid.UUID,
+	expectedSIPCount int,
+	expectedStatus enums.SIPStatus,
+) (*checkResult, error) {
+	// Query all SIPs for this batch.
+	result, err := a.ingestsvc.ListSips(ctx, &goaingest.ListSipsPayload{
+		BatchUUID: ref.New(batchUUID.String()),
+		Limit:     ref.New(50_000),
```

Comment:
Maybe extract this into a file-level constant to make the intent clearer, e.g.:

```go
// maxListSIPsLimit defines the maximum number of SIP records we request when
// polling batch status. This activity currently loads all SIPs into memory,
// but we may switch to a stats-based or aggregated query approach in the future,
// in which case this limit will either go away or be significantly reduced.
const maxListSIPsLimit = 50_000
```

Reply:
I was trying to match [this constant](https://github.com/artefactual-sdps/enduro/blob/main/internal/entfilter/filter.go#L17), I'll reuse it here. Good point about aggregated queries to avoid the limit, that's one of the issues of having some many layers internally, I'd need to add such method to the ingest and the persistence services. Not a huge deal, but something for another PR.

Reply:
Done, and moved to https://github.com/artefactual-sdps/enduro/pull/1458.